src/G2/Equiv/EquivADT.hs

```diff
@@ -22,45 +21,65 @@ import Data.Data
 import Data.Hashable
 
 -- the bool is True if guarded coinduction can be used
+-- TODO do I need all of these typeclasses?
 data Obligation = Ob Bool Expr Expr
                   deriving (Show, Eq, Read, Generic, Typeable, Data)
 
 instance Hashable Obligation
 
-proofObligations :: State t ->
+proofObligations :: (HS.HashSet Name, HS.HashSet Name) ->
+                    State t ->
                     State t ->
                     Expr ->
                     Expr ->
                     Maybe (HS.HashSet Obligation)
-proofObligations s1 s2 e1 e2 =
-  exprPairing s1 s2 e1 e2 HS.empty False
+proofObligations ns_pair s1 s2 e1 e2 =
+  exprPairing ns_pair s1 s2 e1 e2 HS.empty [] [] False
 
-exprPairing :: State t ->
+-- TODO do not allow inlining of variables in ns
+-- TODO have more cases for handling variables that are in ns?
+exprPairing :: (HS.HashSet Name, HS.HashSet Name) ->
+               State t ->
                State t ->
                Expr ->
                Expr ->
                HS.HashSet Obligation ->
+               [Name] ->
+               [Name] ->
                Bool ->
                Maybe (HS.HashSet Obligation)
```

Comment:
Comments explaining at least the `(HS.HashSet Name, HS.HashSet Name)`, `HS.HashSet Obligation`, and two `[Name]` arguments would be helpful.  (For Hackage purposes, argument descriptions can be neatly formatted by using "-- ^ COMMENT HERE" on the same line as the argument, for instance:
https://github.com/BillHallahan/G2/blob/master/src/G2/Data/UFMap.hs#L163-L168
)